GOVUK input for filters in around page

lucascumsille opened the pull request:
Fixes: https://github.com/mysociety/societyworks/issues/3821
Fixes: https://github.com/mysociety/societyworks/issues/4214
Fixes: https://github.com/mysociety/societyworks/issues/4215

IMPORTANT: This update breaks Camden, therefore I think we should tackle [this issue](https://github.com/mysociety/societyworks/issues/3874) before merging this PR. The Alternative I could add a commit here fixing the issue that Camden currently has, which is the `line-height` property, and leave the tidy up ticket in his original order.

### Mobile
<img width="370" alt="Screenshot 2024-05-01 at 07 23 55" src="https://github.com/mysociety/fixmystreet/assets/13790153/ed1d9aa1-8ad2-4ff7-86da-9b6c80e52579">


### Desktop
<img width="1912" alt="Screenshot 2024-05-01 at 07 22 45" src="https://github.com/mysociety/fixmystreet/assets/13790153/34af3a38-bdff-48c0-9dcf-86353ede0591">


---------------

For [this commit ](https://github.com/mysociety/fixmystreet/pull/4728/commits/6a69f8eb942701f31ff75690e6ce4206c97028e4) I replaced `multi-select.min.js` with the one created on [this PR](https://github.com/mysociety/jquery-multi-select/pull/42) on the jQuery-multi-select repo.

- [x] Clean unnecessary code, specially on `_report_list.scss`
- [x] Override the default font size at source when compiling the GOVUK CSS so then it would apply everywhere the same(from 19px to 16px)
- [x] Decide best layout horizontal or vertical stacking
- [x] Test on mobile
- [x] Check Camden
- [x] Update labels so we comply with WCAG 2.53 and fix [issue 4214](https://github.com/mysociety/societyworks/issues/4214)

lucascumsille commented:
@dracos in order to fix https://github.com/mysociety/societyworks/issues/4214
I was thinking to use `aria-labelledby` attribute, we would require to add an `id` for `label` and the fake label `span`(Option 1). I would probably make those changes in the plugin itself. However that would only cover the `label` and `.multi-select-button`, the creation of the fake label, it's something we do in FixMyStreet(I think), which makes me wonder if we could just undo that, basically not creating the new `span` and use the label, without hiding it and add `aria-label` so we can explain better what the select element does(This would be Option 2).

### Option 1 
```
 <label class="hidden-js" id="filter_category_label">About - Select a category</label>
  <span id="filter_category_text">About</span>
  
  <span class="multi-select-button" role="button" aria-haspopup="true" tabindex="0" aria-labelledby="filter_category_label filter_category_text">Everything</span>
```

### Option 2
```
  <label id="filter_category_label" for="filter_category_button">About</label>
  <span class="multi-select-button" id="filter_category_button" role="button" aria-haspopup="true" tabindex="0" aria-labelledby="filter_category_label">Everything</span>
```

dracos commented:
If we're changing the structure to look more like normal fields, it doesn't have to be "About" any more, that's not needed, it's no longer displayed like a readable sentence so doesn't need to be one. It could be "Category". Does that help? Then there's no need for things to be different anywhere. Or have I misunderstood?

dracos commented:
(The history of this FMS changing label to span is https://github.com/mysociety/fixmystreet/pull/3881 and https://github.com/mysociety/fixmystreet/pull/3863 before that.)

lucascumsille commented:
> If we're changing the structure to look more like normal fields, it doesn't have to be "About" any more, that's not needed, it's no longer displayed like a readable sentence so doesn't need to be one. It could be "Category". Does that help? Then there's no need for things to be different anywhere. Or have I misunderstood?

To clarify, In this case we would have `label` + `.multi-select-button` However in terms of markup should we  still include the `for` attribute to the `label` for accessibility purpose. I agree with category it should suffice.

lucascumsille commented:
@dracos I see good to know, I brought the `for` because I didn't see it in this PR, but I can see it on the live website, so I'll check what is breaking that.
In the mean time I have adjusted the `span` so now it says "Category"

lucascumsille commented:
@dracos `for` attribute was being removed from the label because of this commit(https://github.com/mysociety/jquery-multi-select/commit/ed762f7582a73702a29208f197a065cfbaff37f0) on the jQuery-multi-select plugin. I have added a piece of [code here ](https://github.com/mysociety/fixmystreet/pull/4728/commits/cccf1ac705ec5e2e677ae051d0163fc08ebf1519)to bring it back.

lucascumsille commented:
@dracos Here is a preview of the current state. I have updated the screenshot above as well.

<img width="1912" alt="Screenshot 2024-05-01 at 07 22 45" src="https://github.com/mysociety/fixmystreet/assets/13790153/74ce3b2f-bc7d-47ac-9387-c3abb457ce23">
<img width="370" alt="Screenshot 2024-05-01 at 07 23 55" src="https://github.com/mysociety/fixmystreet/assets/13790153/7ccf9ff2-ef60-47db-8a04-7018f72fba7f">

dracos commented:
> for attribute was being removed from the label because of this commit(https://github.com/mysociety/jquery-multi-select/commit/ed762f7582a73702a29208f197a065cfbaff37f0) on the jQuery-multi-select plugin. I have added a piece of [code here ](https://github.com/mysociety/fixmystreet/pull/4728/commits/cccf1ac705ec5e2e677ae051d0163fc08ebf1519)to bring it back.

I was vrery confused, because the `for` is currently present on the live site, and the live jquery-multi-select includes the commit you're referring to. But then I found my commit https://github.com/mysociety/fixmystreet/commit/235931eb12008f0cb5511587f234694281cfd176 where I altered the FMS jquery-multi-select but didn't put this change in the jquery-multi-select plugin :-/ And when this PR updated the plugin for the new stuff there, that disappeared. So that would explain that! I guess your change is a better way to do it if it has the same effect and doesn't alter the plugin.

lucascumsille commented:
I added [this commit](https://github.com/mysociety/fixmystreet/pull/4728/commits/b1c2dbb7f43f91a2ead5ffb50d50861c735386f8) that should help us with this [issue](https://github.com/mysociety/societyworks/issues/4215). I added it here because I though the layout of the current site will change with this PR, therefore it made more sense to include it here, but let me know if you think it shouldn't be here.

dracos commented:
@lucascumsille  I don't think that's what that issue is about - that issue is saying that if you /change/ the filter, it changes automatically without you having to click a Submit, rather than what is happening on page load. The solution is we either add a submit button (meh, guess is easier not to have one, but could do), or add a message saying filters apply automatically (not what /was/ applied).

lucascumsille commented:
@dracos I added a [fixup](https://github.com/mysociety/fixmystreet/pull/4728/commits/95b0ca2cff4f026f756d205df8745a13a74d7cc9) for the message commit. At the beginning I thought it would be nice to let the user know what filters have been applied, but yeah it does overcomplicate the ticket.
Let me know if you are happy with this approach.

dracos commented on templates/web/base/reports/_list-filters.html:
```diff
@@ -27,26 +28,39 @@
-          <input type="submit" name="filter_update" value="[% loc('Go') %]">
-        </p>
+        <fieldset>
+            <div class="multi-select-filter-group">
```
I think this can just be filter-group - there's no difference?

dracos commented on web/cobrands/fixmystreet/fixmystreet.js:
```diff
@@ -1086,7 +1093,7 @@ $.extend(fixmystreet.set_up, {
 
     function update_label(id, str) {
-        $(id).prev('label').addClass('hidden-js').after(function(){ return $('<span>' + this.innerHTML + '</span>'); });
+        $(id).prev('label').addClass('hidden-js').attr('for', id.slice(1)).after(function(){ return $('<span>' + this.innerHTML + '</span>'); });
```
This needs to be prevAll because the filter announcement gets in the way otherwise

dracos reviewed (changes requested):
Looks good, have added fixup for prevAll and to actually have the aria-label on the select match the visible text (which I think was the issue being complained about in 4214). A number of hopefully small issues to deal with

dracos commented on templates/web/base/reports/_list-filters.html:
```diff
@@ -27,26 +28,39 @@
+        <fieldset>
+            <div class="multi-select-filter-group">
+                <div class="report-list-filters">
+                    [% tprintf(loc('<label for="statuses">Report status</label> %s', "Dropdown of all/fixed/etc"), mark_safe(select_status)) %]
```
Have pushed fixup for this

lucascumsille commented on web/cobrands/sass/_multiselect.scss:
```diff
@@ -158,3 +148,124 @@
+        line-height: $govuk-small-input-visible-size;
+    }
+
+    &:has(> input[type="checkbox"]:checked),
```
@dracos I have added this commit that should solve the issue.
https://github.com/mysociety/fixmystreet/pull/4728/commits/0807af665bb667427607e3da9ce4114f11e5da53

lucascumsille commented:
@dracos Thanks for the feedback I have added a fixup for all the issues above with a thumbs up, I think that should be everything. When we are ready I'll tidy up Camden, which most likely will need some amendments with this PR.

dracos commented on templates/web/base/reports/_list-filters.html:
```diff
@@ -27,26 +28,39 @@
+            <div class="multi-select-filter-group">
+                <div class="report-list-filters">
+                    [% tprintf(loc('<label for="statuses">Report status</label> %s', "Dropdown of all/fixed/etc"), mark_safe(select_status)) %]
+                    <input type="submit" name="filter_update" value="[% loc('Go') %]">
```
There are still two go buttons, one under category, one under show older reports; would be good to only have one at the end (so after the show older tickbox, outside the page==around check, I guess?)

lucascumsille replied:
@dracos I added [this commit ](https://github.com/mysociety/fixmystreet/pull/4728/commits/948c6ce17c8a07f862aa6da41af660279d7396a4), I did a minor tweak to the JS, because the markup couldn't target the button anymore, for being outside of `.report-list-filters`

dracos commented on web/cobrands/fixmystreet/fixmystreet.js:
```diff
@@ -1100,6 +1100,35 @@ $.extend(fixmystreet.set_up, {
+            }
+      
+            input.on('change', function() {
+              if (this.checked) {
```
This is fine, but you can also have this as one line with:
`label.toggleClass('govuk-multi-select__label--checked', this.checked);`

dracos commented on templates/web/base/reports/_list-filter-status.html:
```diff
@@ -1,5 +1,6 @@
 [% SET show_all_states = c.cobrand.filter_show_all_states OR (c.user_exists AND (c.user.is_superuser OR c.user.belongs_to_body(body.id))) %]
 
+<div id="filter-announcement" class="screen-reader-only" role="alert">Information: Filters in use. The current view is customized based on selected filters.</div>
```
Should have a loc() around this text string

dracos reviewed (commented):
Looks good :-) Will need a bit of checking that the changes to report lists/ multi selects etc haven't affected anything else, I guess.

lucascumsille commented:
@dracos I found some issues with three cobrands. I couldn't figure out what was causing it, not sure if we are doing something different with them, I couldn't find a template override.

The problem is with "Category" filter, in the markup the `select` element is not present and also not the `multi-select-button`

I tried commenting out this line:
`$('.js-multiple').make_multi();`, but the select element  wouldn't be created.

The Status was is fine

- [ ] highwaysengland
- [ ] isleofwight
- [ ] TFL

lucascumsille commented:
@dracos I checked all the other cobrands and they are working fine on desktop and mobile.
There were a couple of cobrand that needed some[ tweaks though.](https://github.com/mysociety/fixmystreet/pull/4728/commits/ad64f8e6ac82cb34abf948576828539ad6812d80)

Here is a preview of them:
<img width="1038" alt="Screenshot 2024-06-24 at 08 10 23" src="https://github.com/mysociety/fixmystreet/assets/13790153/a3510dd4-e6c5-4341-b657-40c299b2df63">
<img width="1038" alt="Screenshot 2024-06-24 at 08 10 13" src="https://github.com/mysociety/fixmystreet/assets/13790153/3e1ff058-fe4e-43a3-883a-70e776e68142">
<img width="1038" alt="Screenshot 2024-06-24 at 08 40 59" src="https://github.com/mysociety/fixmystreet/assets/13790153/0fd5a6a8-cf5d-42f3-ae53-eafd83dbbce5">
![Screenshot 2024-06-24 at 08 41 17](https://github.com/mysociety/fixmystreet/assets/13790153/ec68c136-8e60-4a1b-9f3e-8978a3bc092d)
<img width="1038" alt="Screenshot 2024-06-24 at 08 06 51" src="https://github.com/mysociety/fixmystreet/assets/13790153/f9ee02c2-c6cf-446c-be73-d96f7c19eb2e">
<img width="1038" alt="Screenshot 2024-06-24 at 08 12 13" src="https://github.com/mysociety/fixmystreet/assets/13790153/7f099e6c-30aa-4af8-ab97-056d13653383">
<img width="1038" alt="Screenshot 2024-06-24 at 08 01 44" src="https://github.com/mysociety/fixmystreet/assets/13790153/faf796c0-7677-4a7e-8d7e-0b06718c2735">
<img width="1038" alt="Screenshot 2024-06-24 at 08 12 02" src="https://github.com/mysociety/fixmystreet/assets/13790153/037c8237-6f94-4487-b84e-bd9cbe6dd434">
<img width="1038" alt="Screenshot 2024-06-24 at 07 59 27" src="https://github.com/mysociety/fixmystreet/assets/13790153/061fbdf6-3ba6-4731-857a-e32381f0b505">
<img width="1038" alt="Screenshot 2024-06-24 at 08 11 49" src="https://github.com/mysociety/fixmystreet/assets/13790153/b20d0d19-a497-4268-944e-0023d425c58c">
<img width="1038" alt="Screenshot 2024-06-24 at 07 56 28" src="https://github.com/mysociety/fixmystreet/assets/13790153/c8da894a-fabc-4efc-9e48-741328e40719">
<img width="1038" alt="Screenshot 2024-06-24 at 08 11 38" src="https://github.com/mysociety/fixmystreet/assets/13790153/c47298e1-1848-4419-b65c-25ff99ff133d">
<img width="522" alt="Screenshot 2024-06-24 at 07 55 42" src="https://github.com/mysociety/fixmystreet/assets/13790153/73946bf5-49e5-4ed1-bdba-4c8a572396a1">
<img width="1038" alt="Screenshot 2024-06-24 at 08 11 28" src="https://github.com/mysociety/fixmystreet/assets/13790153/6305a617-f14b-45de-ad03-443bb6b6c7e5">
<img width="1038" alt="Screenshot 2024-06-24 at 08 42 50" src="https://github.com/mysociety/fixmystreet/assets/13790153/a5d429bd-7682-40a5-90c1-cab75db948a1">
<img width="1038" alt="Screenshot 2024-06-24 at 08 43 02" src="https://github.com/mysociety/fixmystreet/assets/13790153/543d7066-530a-4efa-baf9-6f78ec6a8c21">
<img width="1038" alt="Screenshot 2024-06-24 at 08 43 11" src="https://github.com/mysociety/fixmystreet/assets/13790153/3952d4ae-52e2-4e2a-ae02-d8b98d9a3ea4">
<img width="1038" alt="Screenshot 2024-06-24 at 08 43 20" src="https://github.com/mysociety/fixmystreet/assets/13790153/a1f7ba29-e327-4f16-ad34-b42b84f133f4">
<img width="1038" alt="Screenshot 2024-06-24 at 08 43 35" src="https://github.com/mysociety/fixmystreet/assets/13790153/02661491-35fe-4157-918d-0f4940be9d1b">
<img width="1038" alt="Screenshot 2024-06-24 at 09 26 29" src="https://github.com/mysociety/fixmystreet/assets/13790153/6230752c-2e8a-408e-9931-3a49dce96f67">
<img width="1038" alt="Screenshot 2024-06-24 at 09 28 04" src="https://github.com/mysociety/fixmystreet/assets/13790153/8e470710-d8ea-43df-9322-011e41f2075e">
<img width="522" alt="Screenshot 2024-06-24 at 07 12 12" src="https://github.com/mysociety/fixmystreet/assets/13790153/06f81875-00c1-4bb9-a1b4-9db796c5ffb0">
<img width="1038" alt="Screenshot 2024-06-24 at 08 11 17" src="https://github.com/mysociety/fixmystreet/assets/13790153/86347c82-9aab-4dec-80d0-95fbd5d3853a">
<img width="1038" alt="Screenshot 2024-06-24 at 08 42 31" src="https://github.com/mysociety/fixmystreet/assets/13790153/9568af7d-8bff-41ac-97c1-391057b0cf53">
<img width="1475" alt="Bexley" src="https://github.com/mysociety/fixmystreet/assets/13790153/9063d721-e6a9-4f57-9482-96bbd12ed314">
<img width="1038" alt="Screenshot 2024-06-24 at 08 11 08" src="https://github.com/mysociety/fixmystreet/assets/13790153/d3b83873-4558-480f-adc4-a0007f0c641a">
<img width="1038" alt="Screenshot 2024-06-24 at 08 42 17" src="https://github.com/mysociety/fixmystreet/assets/13790153/48bfb507-c7e3-468a-8e9b-47e11b651829">
<img width="1038" alt="Screenshot 2024-06-24 at 09 28 47" src="https://github.com/mysociety/fixmystreet/assets/13790153/ac54f0bb-d987-49ac-9ae2-072b9e3ea328">
<img width="1038" alt="Screenshot 2024-06-24 at 09 28 56" src="https://github.com/mysociety/fixmystreet/assets/13790153/d8360d01-037a-4d37-9d27-31d719d0b160">
<img width="1038" alt="Screenshot 2024-06-24 at 09 29 13" src="https://github.com/mysociety/fixmystreet/assets/13790153/ee910b87-b557-4794-bdb3-0b84fbeb9570">
<img width="1038" alt="Screenshot 2024-06-24 at 09 29 21" src="https://github.com/mysociety/fixmystreet/assets/13790153/4bb3a5f5-7f3c-4846-8db5-fe77ee78e379">
<img width="1475" alt="Bathnes" src="https://github.com/mysociety/fixmystreet/assets/13790153/ccd48eea-1c01-419c-8fd6-152baa2f5e93">
<img width="1038" alt="Screenshot 2024-06-24 at 08 42 11" src="https://github.com/mysociety/fixmystreet/assets/13790153/e5fc4115-b835-431d-98c5-8aae4de567ca">
<img width="1038" alt="Screenshot 2024-06-24 at 09 28 38" src="https://github.com/mysociety/fixmystreet/assets/13790153/181d9c3e-3aab-4ac5-986a-1cf63d50bbef">
<img width="1038" alt="Screenshot 2024-06-24 at 08 42 01" src="https://github.com/mysociety/fixmystreet/assets/13790153/fa2b1108-5f9c-4ac6-9765-194225791c2e">
<img width="1038" alt="Screenshot 2024-06-24 at 09 28 27" src="https://github.com/mysociety/fixmystreet/assets/13790153/1edff50d-3ab7-4b1b-963a-38d48b591560">

dracos commented:
Hey, three cobrands - if you haven't set up bodies for them, then it'll just say "Everything" like this:
<img width="378" alt="image" src="https://github.com/mysociety/fixmystreet/assets/154364/55b6bbfe-234e-4a6c-9359-53825c44c4af">
Is that what you mean? That'll be fine in the real thing if so.

lucascumsille commented:
@dracos ohh perfect
> Hey, three cobrands - if you haven't set up bodies for them, then it'll just say "Everything" like this:

Then we should be ready.

There is one thing that might be odd, I didn't see it till I went to admin, with this deployment we only have styled the multi-select classes as GOVUK ones, which was fine in the user side of things because I couldn't see an scenario where NON multi-select `select` elements would be alongside multi-select elements. The admin side of things will have case that look like this:

<img width="1056" alt="Screenshot 2024-06-25 at 10 27 18" src="https://github.com/mysociety/fixmystreet/assets/13790153/ea64ffbe-ea43-46d4-bdce-c8089767ff21">

We currently have [this ticket to tackled the admin side](https://github.com/mysociety/societyworks/issues/3828) which most likely will be next. However I wouldn't mind to add a commit here that changes the style of `select` elements so they all look like that, no matter which page you are.

dracos commented:
I don't think that matters really - that was already the case (if slightly more obvious now). What does the dashboard look like? Currently it's eg 
![image](https://github.com/mysociety/fixmystreet/assets/154364/6a8d8808-14e7-4e6c-aa38-d279da31f9dc)
That's the only other place I can think of.

lucascumsille commented:
me neither happy to deploy and then tackle all those cases with the rest of tickets. @dracos, let me know if you are happy with this and I'll rebase and merge. 
Regarding the commits, I'm thinking of getting rid of the --fixups and squash ["Added styling fix for cobrands"](https://github.com/mysociety/fixmystreet/pull/4728/commits/ad64f8e6ac82cb34abf948576828539ad6812d80) into [Added GOVUK styling to multiselect plugin](https://github.com/mysociety/fixmystreet/pull/4728/commits/d014ab2be5288d4c3c52151304e6f8d2125d6a7c).

But let me know if you'd rather have less commits for this PR.

dracos commented:
@lucascumsille Doing something about the dashboard as I mentioned in my previous message would be the only thing I'd like to be looked at - it looks a bit odd now:
<img width="998" alt="image" src="https://github.com/mysociety/fixmystreet/assets/154364/873ec877-d73a-4c96-ab04-3e7ace2be7c8">
Can we for now just shrink them there or something like that?

lucascumsille commented:
@dracos Sorry I had misread your comment. I added a [commit](https://github.com/mysociety/fixmystreet/pull/4728/commits/8dca6c02d5e4f07d92c7317606b2fbbd98df5254) here that reduces the height of the GOVUK ones only for the dashboard. I did't match it exactly, because I would have to reduce the font-size as well, but this approach should get rid of the oddness.

<img width="970" alt="Screenshot 2024-07-02 at 14 45 59" src="https://github.com/mysociety/fixmystreet/assets/13790153/37e3cd9d-9192-45ba-84e8-6813e0ba65e0">

Let me know if you are happy with it and how you'd like me to handle the rebase. I proposed this a few days ago:
>Regarding the commits, I'm thinking of getting rid of the --fixups and squash ["Added styling fix for cobrands"](https://github.com/mysociety/fixmystreet/pull/4728/commits/ad64f8e6ac82cb34abf948576828539ad6812d80) into [Added GOVUK styling to multiselect plugin](https://github.com/mysociety/fixmystreet/pull/4728/commits/d014ab2be5288d4c3c52151304e6f8d2125d6a7c).

lucascumsille commented:
Thanks @dracos looks good to me, thanks for adding it to the changelog as well =)

dracos commented on web/cobrands/fixmystreet/fixmystreet.js:
```diff
@@ -132,6 +132,13 @@ function isR2L() {
+            settings.menuItemsHTML = '<div class="govuk-multi-select govuk-multi-select--checkboxes">';
+            settings.menuItemHTML = '<label class="govuk-multi-select__label">';
+            settings.menuFieldsetHTML = '<fieldset class="multi-select-fieldset govuk-fieldset">';
+            settings.menuFieldsetLegendHTML = '<fieldset class="multi-select-fieldset govuk-fieldset__legend govuk-fieldset__legend--s">';
```
Not sure how I didn't notice this before, sorry, but this should presumably be a `legend` not a `fieldset`.

lucascumsille commented:
Thanks @dracos well spotted. I create a PR here:
https://github.com/mysociety/fixmystreet/pull/5069
